src/macaron/repo_finder/repo_finder_java.py

```diff
@@ -0,0 +1,327 @@
+# Copyright (c) 2023 - 2023, Oracle and/or its affiliates. All rights reserved.
+# Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl/.
+
+"""This module contains the JavaRepoFinder class to be used for finding Java repositories."""
+import logging
+import re
+from collections.abc import Iterator
+from xml.etree.ElementTree import Element  # nosec
+
+import defusedxml.ElementTree
+from defusedxml.ElementTree import fromstring
+from packageurl import PackageURL
+from requests.exceptions import ReadTimeout
+
+from macaron.config.defaults import defaults
+from macaron.repo_finder.repo_finder_base import BaseRepoFinder
+from macaron.util import send_get_http_raw
+
+logger: logging.Logger = logging.getLogger(__name__)
+
+
+class JavaRepoFinder(BaseRepoFinder):
+    """This class is used to find Java repositories."""
+
+    def __init__(self) -> None:
+        """Initialise the Java repository finder instance."""
+        self.pom_element: Element | None = None
+
+    def find_repo(self, purl: PackageURL) -> Iterator[Iterator[str]]:
+        """
+        Generate iterator from _find_repo that attempts to retrieve a repository URL that matches the passed artifact.
+
+        Parameters
+        ----------
+        purl : PackageURL
+            The PURL of an artifact.
+
+        Yields
+        ------
+        Iterator[Iterator[str]] :
+            The URLs found for the passed artifact.
+        """
+        yield from iter(self._find_repo(purl))  # type: ignore[misc]
```

Comment:
Shouldn't this just be:
```suggestion
        yield from self._find_repo(purl)
```
And the return type would just be `Iterator[str]`? The return type of `Iterator[Iterator[str]]` seems quite odd and I don't see how it is necessary.
I also have the similar comment for the `DepsDevRepoFinder::find_repo()` method.

Reply:
The reason it is that way is due to how finding repositories for Java works. Retrieving a POM and analysing it potentially returns multiple URLs, as a list that is iterable. At that stage the list needs to be iterated in search of valid URLs, and if none are found, the search continues with the next parent POM. Using nested iterators with `yield from` allow this process to continue from where it left off, thereby retrieving the next iterable list, and so on.
While deps.dev does not need this, other future finders possibly might. For example, one that checks multiple sources of knowledge in sequence.
Sadly, `type: ignore[misc]` is needed because `mypy` cannot handle this situation at all. The hints it gives are wrong.

Reply:
Hmm, right. I think I see what you are trying to do.
> Retrieving a POM and analysing it potentially returns multiple URLs, as a list that is iterable. At that stage the list needs to be iterated in search of valid URLs, and if none are found, the search continues with the next parent POM.

Having said that, shouldn't this logic be internal to only the `JavaRepoFinder`? I don't quite see the point of iterating through multiple calls of `repo_finder.find_repo` until a valid result is found, as opposed to getting a valid result from a single call. Interacting until a valid URL is found, imo, should happen within the `JavaRepoFinder` class, not outside of it.

Reply:
I may have got carried away with the idea of iterators inside iterators. The simpler solution _is_ to move the URL validation function (again) so it can be used inside the Repo Finders.